src/test/java/engineering/swat/watch/TortureTests.java

```diff
@@ -0,0 +1,387 @@
+            await("We should see only exactly the " + amountOfWatchersActive + " events we expect")
+                .failFast(() -> !exceptions.isEmpty())
+                .pollDelay(TestHelper.NORMAL_WAIT.minusMillis(100))
+                .until(seen::size, Predicate.isEqual(amountOfWatchersActive))
+                ;
+            if (!exceptions.isEmpty()) {
+                fail(exceptions.pop());
+            }
+        }
+        finally {
+            stopAll.release(amountOfWatchersActive);
+        }
+
+    }
+
+
+
+    @Test
+    //Deletes can race the filesystem, so you might miss a few files in a dir, if that dir is already deleted
+    @EnabledIfEnvironmentVariable(named="TORTURE_DELETE", matches="true")
+    void pressureOnFSShouldNotMissDeletes() throws InterruptedException, IOException {
+        final var root = testDir.getTestDirectory();
+        var pool = Executors.newCachedThreadPool();
+
+        Set<Path> pathsWritten;
+        var seenDeletes = ConcurrentHashMap.<Path>newKeySet();
+        var io = new IOGenerator(THREADS, root, pool);
+        try {
+            io.start();
+            Thread.sleep(TestHelper.NORMAL_WAIT.toMillis());
+            pathsWritten = io.stop();
+
+            final var events = new AtomicInteger(0);
+            final var happened = new Semaphore(0);
+            var watchConfig = Watcher.watch(testDir.getTestDirectory(), WatchScope.PATH_AND_ALL_DESCENDANTS)
+                .withExecutor(pool)
+                .onEvent(ev -> {
+                    events.getAndIncrement();
+                    happened.release();
+                    var fullPath = ev.calculateFullPath();
+                    switch (ev.getKind()) {
+                        case DELETED:
+                            seenDeletes.add(fullPath);
+                            break;
+                        case MODIFIED:
+                            // happens on dir level, as the files are getting removed
+                            break;
+                        default:
+                            logger.error("Unexpected event: {}", ev);
+                            break;
+                    }
+                });
+
+            try (var activeWatch = watchConfig.start() ) {
+                logger.info("Deleting files now", THREADS);
+                testDir.deleteAllFiles();
+                logger.info("Waiting for the events processing to stabilize");
+                waitForStable(events, happened);
+            }
+        }
+        finally {
+            try {
+                io.stop();
+            }
+            catch (Throwable _ignored) {}
+            // shutdown the pool (so no new events are registered)
+            pool.shutdown();
+        }
+
+        // but wait till all scheduled tasks have been completed
+        pool.awaitTermination(10, TimeUnit.SECONDS);
+
+        logger.info("Comparing events and files seen");
+        // now make sure that the two sets are the same
```

Comment:
Nit: The loop below only checks that `pathsWritten` is a subset of `seenDeletes`, but not the other way around

Reply:
well, you might get deletes for the directories during the cleanup, that is why.